File: Lucy/2025/3-W02/236_Lowest_Common_Ancestor_of_a_Binary_Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

// 두번째 풀이
function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
  if (!root || root === p || root === q) return root;

  const left = lowestCommonAncestor(root.left, p, q);
  const right = lowestCommonAncestor(root.right, p, q);

  if (left && right) return root;

  return left || right;
}

// 첫 번째 풀이
function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
  const dfs = (node: TreeNode | null, target: TreeNode | null, ancestors: TreeNode[]): boolean => {
    if (node === null) return false;

    ancestors.push(node);

    if (node.val === target.val) return true;

    if (dfs(node.left, target, ancestors) || dfs(node.right, target, ancestors)) return true;

    ancestors.pop();

    return false;
  };

  const pAncestors: TreeNode[] = [];
  const qAncestors: TreeNode[] = [];

  dfs(root, p, pAncestors);
  dfs(root, q, qAncestors);

  const commonAncestor: TreeNode | null = pAncestors.filter((ancestor) => qAncestors.includes(ancestor)).at(-1);

  return commonAncestor;
}

File: Seori/LeetCode 75/Binary Tree_DFS/236. Lowest Common Ancestor of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # 재귀 종료조건
        if root is None:
            return root
        
        # 재귀 종료조건 : p나 q를 발견한 경우
        if root == p or root == q:
            return root
        
        # p, q를 발견할 때까지 재귀 탐색
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        
        # p, q를 모두 발견한 경우, 현재 노드가 LCA
        if left != None and right != None:
            return root
        # p, q 중 하나만 발견한 경우, 해당 노드가 LCA
        elif left != None:
            return left
        elif right != None:
            return right